sklearn/decomposition/nmf.py

```diff
@@ -254,22 +344,287 @@ def _nls_subproblem(V, W, H, tol, max_iter, sigma=0.01, beta=0.1):
+    """Compute Non-negative Matrix Factorization (NMF) with Coordinate Descent
+
+    Find (W, H) that minimize || X - W.H ||_2
+    with an alternative least square minimization of W and H.
+    Each minimization is done with a Cyclic Coordinate Descent.
 
     Parameters
     ----------
-    n_components : int or None
-        Number of components, if n_components is not set all components
-        are kept
+    X : array-like, shape (n_samples, n_features)
+        Constant matrix.
+
+    W : array-like, shape (n_samples, n_components)
+        Initial guess for the solution.
+
+    H : array-like, shape (n_components, n_features)
+        Initial guess for the solution.
+
+    tol : float, default: 1e-4
+        Tolerance of the stopping condition.
+
+    max_iter : integer, default: 200
+        Maximum number of iterations before timing out.
+
+    alpha : double, default: 0.
+        Constant that multiplies the regularization terms.
+
+    l1_ratio : double, default: 0.
+        The regularization mixing parameter, with 0 <= l1_ratio <= 1.
+        For l1_ratio = 0 the penalty is an L2 penalty.
+        For l1_ratio = 1 it is an L1 penalty.
+        For 0 < l1_ratio < 1, the penalty is a combination of L1 and L2.
+
+    regularization : 'both' | 'components' | 'transformation' | None
+        Select whether the regularization affects the components (H), the
+        transformation (W), both or none of them.
+
+    update_H : boolean, default: True
+        Set to True, both W and H will be estimated from initial guesses.
+        Set to False, only W will be estimated.
+
+    verbose : integer, default: 0
+        The verbosity level.
+
+    shuffle : boolean, default: False
+        If True, the samples will be taken in shuffled order during
+        coordinate descent.
+
+    random_state : integer seed, RandomState instance, or None (default)
+        Random number generator seed control.
+
+    Returns
+    -------
+    W : array-like, shape (n_samples, n_components)
+        Solution to the non-negative least squares problem.
+
+    H : array-like, shape (n_components, n_features)
+        Solution to the non-negative least squares problem.
+
+    n_iter : int
+        The number of iterations done by the algorithm.
 
-    init :  'nndsvd' |  'nndsvda' | 'nndsvdar' | 'random'
+    References
+    ----------
+    Hsieh, Cho-Jui, and Inderjit S. Dhillon. "Fast coordinate descent methods
+    with variable selection for non-negative matrix factorization."
+    Proceedings of the 17th ACM SIGKDD. ACM, 2011.
+    """
+    # so W and Ht are both in C order in memory
+    Ht = check_array(H.T, order='C')
+    X = check_array(X, accept_sparse='csr')
+
+    alpha_H = 0.
+    alpha_W = 0.
+    if regularization in ('both', 'components'):
+        alpha_H = alpha
+    if regularization in ('both', 'transformation'):
+        alpha_W = alpha
+
+    rng = check_random_state(random_state)
+
+    for n_iter in range(max_iter):
+            v = 0.
```

Comment:
one letter variable name could be clearer. Maybe call it `violation` and explain in the comments what it means.